src/audio/sink_source_utils.c

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/sink_source_utils.h>
+#include <sof/audio/sink_api.h>
+#include <sof/audio/source_api.h>
+#include <sof/common.h>
+#include <rtos/string.h>
+#include <sof/math/numbers.h>
+#include <assert.h>
+
+int source_to_sink_copy(struct source __sparse_cache *source, struct sink __sparse_cache *sink,
+			bool free, size_t size)
+{
+	uint8_t __sparse_cache *src_ptr;
+	uint8_t __sparse_cache *src_begin;
+	uint8_t __sparse_cache *src_end;
+	uint8_t __sparse_cache *dst_ptr;
+	uint8_t __sparse_cache *dst_begin;
+	uint8_t __sparse_cache *dst_end;
```

Comment:
I think that's too much now... Currently we use `__sparse_cache` for `struct comp_buffer` and since `struct audio_stream` since it's embedded into it. But we don't mark buffer data as `__sparse_cache` even though we allocate it as cached (`rballoc()` always returns cached addresses). So I think most of your added annotations aren't needed.